spec/unit/functions/index_spec.rb

```diff
@@ -0,0 +1,175 @@
+        file { "$n": ensure => present }
+      MANIFEST
+      expect(catalog.resource(:file, "6")['ensure']).to eq('present')
+    end
+
+    it 'a given value returns index of first found substring given as a regexp' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $s = "foobarfuub"
+        $n = $s.index(/f(oo|uu)b/)
+        file { "$n": ensure => present }
+      MANIFEST
+      expect(catalog.resource(:file, "0")['ensure']).to eq('present')
+    end
+  end
+
+  context "should be callable on an iterable" do
+    it 'for example a reverse_each' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $a = [1,2,3]
+        $n = $a.reverse_each.index |$v| { $v == 1 }
+        file { "$n": ensure => present }
+      MANIFEST
+      expect(catalog.resource(:file, "2")['ensure']).to eq('present')
+    end
+  end
+
+  context 'stops iteration when result is known' do
+    it 'true when boolean true is found' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $a = [1,2,3]
+        $n = $a.index |$i, $v| { if $i == 0 { true } else { fail("unwanted") } }
+        file { "$n": ensure => present }
+      MANIFEST
+
+      expect(catalog.resource(:file, "0")['ensure']).to eq('present')
+    end
+  end
+
+  context 'returns undef when value is not found' do
+    it 'when using array and a lambda' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $a = [1,2,3]
+        $n = $a.index |$v| { $v == 'blue' }
+        file { "test": ensure => if $n =~ Undef { present } else {absent} }
+      MANIFEST
+      expect(catalog.resource(:file, "test")['ensure']).to eq('present')
+    end
+
+    it 'when using array and a value' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $a = [1,2,3]
+        $n = $a.index('blue')
+        file { "test": ensure => if $n =~ Undef { present } else {absent} }
+      MANIFEST
+      expect(catalog.resource(:file, "test")['ensure']).to eq('present')
+    end
+
+    it 'when using a hash and a lambda' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $h = {1 => 10, 2 => 20, 3 => 30}
+        $n = $h.index |$v| { $v == 'blue' }
+        file { "test": ensure => if $n =~ Undef { present } else {absent} }
+      MANIFEST
+      expect(catalog.resource(:file, "test")['ensure']).to eq('present')
+    end
+
+    it 'when using a hash and a value' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $h = {1 => 10, 2 => 20, 3 => 30}
+        $n = $h.index('blue')
+        file { "test": ensure => if $n =~ Undef { present } else {absent} }
+      MANIFEST
+      expect(catalog.resource(:file, "test")['ensure']).to eq('present')
+    end
+
+    it 'when using a String and a value' do
+      catalog = compile_to_catalog(<<-MANIFEST)
+        $s = "foobarfuub"
+        $n = $s.index('banana')
+        file { "test": ensure => if $n =~ Undef { present } else {absent} }
+      MANIFEST
+      expect(catalog.resource(:file, "test")['ensure']).to eq('present')
+    end
```

Comment:
I would kind of like to also see a test for a String and lambda, even though I think under the hood that is kind of covered by the Array test.

Reply:
Lines 69-76 has that already ;-)